File: middleware/queue_manager.py
import pika
import logging
import json
from protocol.messages import MESSAGE_TYPE_BATCH, BatchMessage, DatasetType, _create_record_from_string
from ..common.config import get_config
from middleware.middleware import (
    MessageMiddleware,
    MessageMiddlewareDisconnectedError,
    MessageMiddlewareMessageError,
    MessageMiddlewareCloseError,
    MessageMiddlewareDeleteError,
)


class QueueManager(MessageMiddleware):
    """Manages RabbitMQ connections and queue operations for the filter node"""

    def __init__(
        self, host=None, queue_name=None, port=None, username=None, password=None
    ):
        # Load configuration
        config = get_config()
        rabbitmq_config = config.get_rabbitmq_config()
        filter_config = config.get_filter_config()

        # Use provided parameters or fall back to configuration
        self.host = host or rabbitmq_config["host"]
        self.port = port or rabbitmq_config["port"]
        self.username = username or rabbitmq_config["username"]
        self.password = password or rabbitmq_config["password"]
        self.connection = None
        self.channel = None
        self._consuming = False
        self._consumer_tag = None

        # Configure input queue and output exchanges from config file
        self.input_queue = filter_config["input_queue"]
        self.transactions_exchange = filter_config["transactions_exchange"]
        self.transaction_items_exchange = filter_config["transaction_items_exchange"]

        # Create input queue mapping
        self.input_queue_mapping = {
            DatasetType.TRANSACTIONS: self.input_queue,
            DatasetType.TRANSACTION_ITEMS: self.input_queue,
        }

        logging.info(
            f"action: queue_manager_init | input_queue: {self.input_queue} | "
            f"transactions_exchange: {self.transactions_exchange} | "
            f"transaction_items_exchange: {self.transaction_items_exchange} | "
        )

    def connect(self):
        """Establish connection to RabbitMQ"""
        try:
            credentials = pika.PlainCredentials(self.username, self.password)
            parameters = pika.ConnectionParameters(
                host=self.host, port=self.port, credentials=credentials
            )
            self.connection = pika.BlockingConnection(parameters)
            self.channel = self.connection.channel()

            # Declare input queue
            self.channel.queue_declare(queue=self.input_queue, durable=True)

            # Declare all output exchanges
            if self.transactions_exchange:
                self.channel.exchange_declare(
                    exchange=self.transactions_exchange,
                    exchange_type="direct",  # Direct exchange for specific routing
                    durable=True,
                )
            if self.transaction_items_exchange:
                self.channel.exchange_declare(
                    exchange=self.transaction_items_exchange,
                    exchange_type="direct",  # Direct exchange for specific routing
                    durable=True,
                )

            logging.info(
                f"action: rabbitmq_connect | result: success | host: {self.host} | "
                f"input_queue: {self.input_queue} | transactions_exchange: {self.transactions_exchange} | "
                f"transaction_items_exchange: {self.transaction_items_exchange}"
            )
            return True

        except Exception as e:
            logging.error(f"action: rabbitmq_connect | result: fail | error: {e}")
            return False

    def disconnect(self):
        """Close RabbitMQ connection"""
        try:
            if self.connection and not self.connection.is_closed:
                self.connection.close()
            logging.info("action: rabbitmq_disconnect | result: success")
        except Exception as e:
            logging.error(f"action: rabbitmq_disconnect | result: fail | error: {e}")

    def send_dataset_batch(self, batch_message):
        """Route dataset batch to appropriate input queue based on dataset type (for compatibility)"""
        try:
            queue_name = self.input_queue_mapping.get(batch_message.dataset_type)
            if not queue_name:
                raise ValueError(
                    f"No input queue mapping for dataset type: {batch_message.dataset_type}"
                )

            # Serialize the batch message for the queue
            message_data = {
                "dataset_type": batch_message.dataset_type,
                "records": [record.serialize() for record in batch_message.records],
                "eof": batch_message.eof,
            }

            message_body = json.dumps(message_data)

            self.channel.basic_publish(
                exchange="",
                routing_key=queue_name,
                body=message_body,
                properties=pika.BasicProperties(
                    delivery_mode=2
                ),  # Make message persistent
            )

            logging.info(
                f"action: send_batch | result: success | queue: {queue_name} | "
                f"dataset_type: {batch_message.dataset_type} | "
                f"record_count: {len(batch_message.records)} | eof: {batch_message.eof}"
            )

            return True

        except Exception as e:
            logging.error(f"action: send_batch | result: fail | error: {e}")
            return False

    def start_consuming_transactions(self, callback):
        """Start consuming from configured input queue and call callback for each message"""
        try:
            input_queue = self.input_queue

            def wrapper(ch, method, properties, body):
                try:
                    msg_type = body[0]

                    if msg_type == MESSAGE_TYPE_BATCH:
                        batch_message = BatchMessage.from_data(body)
                    else:
                        raise ValueError(f"Unknown message type: {msg_type}")

                    # Call the callback with the parsed data
                    callback(batch_message)

                    # Acknowledge the message
                    ch.basic_ack(delivery_tag=method.delivery_tag)

                except Exception as e:
                    logging.error(
                        f"action: process_transaction | result: fail | error: {e}"
                    )
                    # Reject and requeue the message
                    ch.basic_nack(delivery_tag=method.delivery_tag, requeue=True)

            self.channel.basic_consume(queue=input_queue, on_message_callback=wrapper)

            logging.info(
                f"action: start_consuming_transactions | result: success | queue: {input_queue}"
            )
            self.channel.start_consuming()

        except Exception as e:
            logging.error(
                f"action: start_consuming_transactions | result: fail | error: {e}"
            )

    def encode_to_byte_array(batch_message):
        # [MessageType][DatasetType][EOF][RecordCount][Records...]
        data = bytearray()
        data.append(MESSAGE_TYPE_BATCH)
        data.append(batch_message.dataset_type)

        # Build content: EOF|RecordCount|Record1|Record2|...
        content = f"{1 if batch_message.eof else 0}|{len(batch_message.records)}"
        for record in batch_message.records:
            content += "|" + record.serialize()

        data.extend(content.encode("utf-8"))

        return data

    def send_filtered_batch(self, exchange_name, batch_message):
        """Send filtered batch to specific output exchange by name"""
        try:
            byte_array = self.encode_to_byte_array(batch_message)

            self.channel.basic_publish(
                exchange=exchange_name,
                routing_key="",  # No specific routing key needed for direct exchange
                body=byte_array,
                properties=pika.BasicProperties(
                    delivery_mode=2
                ),  # Make message persistent
            )

            logging.info(
                f"action: send_filtered_batch | result: success | exchange: {exchange_name} | "
                f"original_type: {batch_message.dataset_type} | "
                f"record_count: {len(batch_message.records)} | eof: {batch_message.eof}"
            )

            return True

        except Exception as e:
            logging.error(f"action: send_filtered_batch | result: fail | error: {e}")
            return False

    def send_to_dataset_output_exchanges(self, batch_message):
        """Send filtered batch to all appropriate output exchanges for a dataset type"""
        if batch_message.dataset_type == DatasetType.TRANSACTIONS:
            exchange = self.transactions_exchange
        elif batch_message.dataset_type == DatasetType.TRANSACTION_ITEMS:
            exchange = self.transaction_items_exchange
        else:
            logging.warning(
                f"action: send_to_dataset_output_exchanges | result: fail | "
                f"error: DATASET TYPE: {batch_message.dataset_type} NOT SUPPORTED"
            )
            return False

        success = self.send_filtered_batch(
            exchange, batch_message
        )
        if success:
            logging.info(
                f"action: send_to_dataset_output_exchanges | result: success | "
                f"dataset_type: {batch_message.dataset_type}"
            )
            return True
        else:
            logging.error(
                f"action: send_to_dataset_output_exchanges | result: partial_fail | "
                f"dataset_type: {batch_message.dataset_type}"
            )
            return False

    def stop_consuming(self):
        """Stop consuming messages"""
        try:
            if self.channel:
                self.channel.stop_consuming()
            logging.info("action: stop_consuming | result: success")
        except Exception as e:
            logging.error(f"action: stop_consuming | result: fail | error: {e}")

    # MessageMiddleware interface methods
    def start_consuming(self, callback):
        """Start consuming messages - delegates to start_consuming_transactions"""
        self.start_consuming_transactions(callback)

    def send(self, message):
        """Send message to default transactions exchange (MessageMiddleware interface)"""
        try:
            self.channel.basic_publish(
                exchange=self.transactions_exchange,  
                routing_key="",
                body=message,
                properties=pika.BasicProperties(
                    delivery_mode=2
                ),  # Make message persistent
            )
            return True
        except Exception as e:
            logging.error(f"action: send | result: fail | error: {e}")
            return False

    def close(self):
        """Close the connection"""
        try:
            if self.connection and not self.connection.is_closed:
                self.connection.close()
            logging.info("action: close | result: success")
        except Exception as e:
            logging.error(f"action: close | result: fail | error: {e}")

    def delete(self, exchange_name=None):
        """Delete an exchange"""
        try:
            if self.channel:
                # Use provided exchange or default to transactions_exchange
                target_exchange = exchange_name or self.transactions_exchange
                self.channel.exchange_delete(exchange=target_exchange)
            logging.info(
                f"action: delete | result: success | exchange: {target_exchange}"
            )
        except Exception as e:
            logging.error(
                f"action: delete | result: fail | exchange: {target_exchange} | error: {e}"
            )

File: middleware/queue_manager.go
package middleware

import (
	"encoding/json"
	"fmt"
	"log"
	"sync"

	"github.com/distribuidos-Coffee-Shop-Analysis/nodes/common"
	"github.com/distribuidos-Coffee-Shop-Analysis/nodes/protocol"
	amqp "github.com/rabbitmq/amqp091-go"
)

const HEARTBEAT_SECONDS = 3

// CleanableHandler defines the interface for handlers that support cleanup operations
type CleanableHandler interface {
	OnCleanup(clientID string)
}

// CleanupMessage represents the JSON structure of cleanup messages
type CleanupMessage struct {
	ClientID string `json:"client_id"`
}

// QueueManager manages RabbitMQ connections and queue operations
type QueueManager struct {
	host       string
	port       int
	username   string
	password   string
	Connection *amqp.Connection
	channel    *amqp.Channel

	// Pipeline components
	pipelines  []*Pipeline
	pipelineMu sync.Mutex

	Wiring *common.NodeWiring
}

func NewQueueManagerWithWiring(w *common.NodeWiring) *QueueManager {
	cfg := common.GetConfig()
	r := cfg.GetRabbitmqConfig()
	return &QueueManager{
		host:      r.Host,
		port:      r.Port,
		username:  r.Username,
		password:  r.Password,
		pipelines: make([]*Pipeline, 0),
		Wiring:    w,
	}
}

func (qm *QueueManager) Connect() error {
	var err error

	connStr := fmt.Sprintf("amqp://%s:%s@%s:%d/?heartbeat=%d",
		qm.username, qm.password, qm.host, qm.port, HEARTBEAT_SECONDS)

	qm.Connection, err = amqp.Dial(connStr)
	if err != nil {
		log.Printf("action: rabbitmq_connect | result: fail | error: %v", err)
		return err
	}

	qm.channel, err = qm.Connection.Channel()
	if err != nil {
		log.Printf("action: rabbitmq_channel | result: fail | error: %v", err)
		return err
	}

	// exchanges
	for _, ex := range qm.Wiring.DeclareExchs {
		kind := "direct"
		if err := qm.channel.ExchangeDeclare(ex, kind, true, false, false, false, nil); err != nil {
			_ = qm.channel.Close()
			_ = qm.Connection.Close()
			return fmt.Errorf("declare exchange %s: %w", ex, err)
		}
	}

	// Declare and bind queues for each binding
	for i, b := range qm.Wiring.Bindings {
		var queueName string

		if b.UseSharedQueue {
			queueName = qm.Wiring.SharedQueueName
		} else {
			if len(qm.Wiring.Bindings) > 1 {
				queueName = fmt.Sprintf("%s_%d", qm.Wiring.IndividualQueueName, i)
			} else {
				queueName = qm.Wiring.IndividualQueueName
			}
		}

		q, err := qm.channel.QueueDeclare(queueName, true, false, false, false, nil)
		if err != nil {
			_ = qm.channel.Close()
			_ = qm.Connection.Close()
			return fmt.Errorf("declare queue %s: %w", queueName, err)
		}

		if err := qm.channel.QueueBind(q.Name, b.RoutingKey, b.Exchange, false, nil); err != nil {
			_ = qm.channel.Close()
			_ = qm.Connection.Close()
			return fmt.Errorf("bind queue %s to exchange %s: %w", q.Name, b.Exchange, err)
		}

		log.Printf("action: queue_setup | queue: %s | exchange: %s | routing_key: %s | binding_index: %d | shared: %v",
			queueName, b.Exchange, b.RoutingKey, i, b.UseSharedQueue)
	}

	return nil
}

func (qm *QueueManager) Disconnect() {
	if qm.Connection != nil && !qm.Connection.IsClosed() {
		qm.Connection.Close()
	}
	log.Println("action: rabbitmq_disconnect | result: success")
}

func (qm *QueueManager) getQueueName(bindingIndex int) string {
	b := qm.Wiring.Bindings[bindingIndex]

	if b.UseSharedQueue {
		return qm.Wiring.SharedQueueName
	}

	if len(qm.Wiring.Bindings) > 1 {
		return fmt.Sprintf("%s_%d", qm.Wiring.IndividualQueueName, bindingIndex)
	}

	return qm.Wiring.IndividualQueueName
}

// StartConsuming starts consuming from configured input queues.
func (qm *QueueManager) StartConsuming(callback func(batch *protocol.BatchMessage, delivery amqp.Delivery)) error {
	cfg := common.GetConfig()
	pipelineConfig := cfg.GetPipelineConfig(qm.Wiring.Role)

	log.Printf("action: start_consuming_pipeline | role: %s | consumers: %d | prefetch: %d | processors: %d | num_queues: %d",
		qm.Wiring.Role,
		pipelineConfig.NumConsumers,
		pipelineConfig.ConsumerPrefetch,
		pipelineConfig.NumProcessors,
		len(qm.Wiring.Bindings))

	var wg sync.WaitGroup

	// Start a pipeline for each queue
	for i := range qm.Wiring.Bindings {
		queueName := qm.getQueueName(i)

		config := PipelineConfig{
			NumConsumers:     pipelineConfig.NumConsumers,
			ConsumerPrefetch: pipelineConfig.ConsumerPrefetch,
			NumProcessors:    pipelineConfig.NumProcessors,
			InputBufferSize:  pipelineConfig.InputBufferSize,
		}

		pipeline := NewPipeline(config, qm.Connection)

		qm.pipelineMu.Lock()
		qm.pipelines = append(qm.pipelines, pipeline)
		qm.pipelineMu.Unlock()

		if err := pipeline.Start(queueName, callback); err != nil {
			qm.StopConsuming()
			return fmt.Errorf("failed to start pipeline for queue %s: %w", queueName, err)
		}

		wg.Add(1)
		go func(p *Pipeline) {
			defer wg.Done()
			p.Wait()
		}(pipeline)
	}

	wg.Wait()

	return nil
}

// StopConsuming stops all pipelines.
func (qm *QueueManager) StopConsuming() {
	log.Println("action: stop_consuming | status: stopping_pipelines")

	qm.pipelineMu.Lock()
	pipelines := qm.pipelines
	qm.pipelineMu.Unlock()

	for _, p := range pipelines {
		p.Stop()
	}

	log.Println("action: stop_consuming | result: success")
}

// Close closes the connection and stops all pipelines.
func (qm *QueueManager) Close() error {
	qm.StopConsuming()

	var err error
	if qm.Connection != nil && !qm.Connection.IsClosed() {
		err = qm.Connection.Close()
	}
	if err != nil {
		log.Printf("action: close | result: fail | error: %v", err)
	} else {
		log.Println("action: close | result: success")
	}
	return err
}

// StartCleanupListener starts a goroutine that listens for cleanup signals
// and calls the handler's OnCleanup method for each client that needs cleanup
func (qm *QueueManager) StartCleanupListener(handler CleanableHandler) error {
	// Declare cleanup exchange (fanout, durable)
	err := qm.channel.ExchangeDeclare(
		"cleanup_exchange", // name
		"fanout",           // type
		true,               // durable
		false,              // auto-deleted
		false,              // internal
		false,              // no-wait
		nil,                // arguments
	)
	if err != nil {
		log.Printf("action: cleanup_exchange_declare | result: fail | error: %v", err)
		return fmt.Errorf("failed to declare cleanup exchange: %w", err)
	}

	// Declare an exclusive, auto-delete queue for this node
	queue, err := qm.channel.QueueDeclare(
		"",    // name (empty = server generates unique name)
		false, // durable (false for exclusive queues)
		true,  // auto-delete (deleted when consumer disconnects)
		true,  // exclusive (only this connection can access)
		false, // no-wait
		nil,   // arguments
	)
	if err != nil {
		log.Printf("action: cleanup_queue_declare | result: fail | error: %v", err)
		return fmt.Errorf("failed to declare cleanup queue: %w", err)
	}

	// Bind queue to cleanup exchange
	err = qm.channel.QueueBind(
		queue.Name,         // queue name
		"",                 // routing key (ignored for fanout)
		"cleanup_exchange", // exchange
		false,              // no-wait
		nil,                // arguments
	)
	if err != nil {
		log.Printf("action: cleanup_queue_bind | result: fail | error: %v", err)
		return fmt.Errorf("failed to bind cleanup queue: %w", err)
	}

	log.Printf("action: cleanup_listener_setup | result: success | queue: %s", queue.Name)

	// Start consuming from cleanup queue
	msgs, err := qm.channel.Consume(
		queue.Name, // queue
		"",         // consumer tag (empty = server generates)
		false,      // auto-ack
		true,       // exclusive
		false,      // no-local
		false,      // no-wait
		nil,        // arguments
	)
	if err != nil {
		log.Printf("action: cleanup_consume | result: fail | error: %v", err)
		return fmt.Errorf("failed to start consuming cleanup messages: %w", err)
	}

	// Start goroutine to process cleanup messages
	go func() {
		log.Println("action: cleanup_listener_start | result: success | msg: listening for cleanup signals")

		for msg := range msgs {
			var cleanupMsg CleanupMessage
			if err := json.Unmarshal(msg.Body, &cleanupMsg); err != nil {
				log.Printf("action: cleanup_message_decode | result: fail | error: %v", err)
				msg.Ack(false)
				continue
			}

			log.Printf("action: cleanup_received | result: processing | client_id: %s", cleanupMsg.ClientID)

			func() {
				defer func() {
					if r := recover(); r != nil {
						log.Printf("action: cleanup_panic | client_id: %s | error: %v", cleanupMsg.ClientID, r)
						msg.Nack(false, true)
					}
				}()

				handler.OnCleanup(cleanupMsg.ClientID)

				if err := msg.Ack(false); err != nil {
					log.Printf("action: cleanup_ack | client_id: %s | result: fail | error: %v", cleanupMsg.ClientID, err)
				} else {
					log.Printf("action: cleanup_processed | result: success | client_id: %s", cleanupMsg.ClientID)
				}
			}()
		}

		log.Println("action: cleanup_listener_end | result: success | msg: cleanup listener stopped")
	}()

	return nil
}
